### PR TITLE
*: Fix #254

### LINKED ISSUE #254: Alter table error
```
CREATE TABLE `users` (`name` varchar(255),`age` bigint);
INSERT INTO `users` (`name`,`age`) VALUES ("name",2);
ALTER TABLE `users` ADD `weight` bigint;
SELECT * FROM `users`;
[error] Error: key not exist
```

### PR DESCRIPTION
**Pull request by zimulala**
Fix https://github.com/pingcap/tidb/issues/254

### CODE REVIEW

File: ddl/ddl.go [hunk 1]
```diff
@@ -489,9 +490,9 @@ func (d *ddl) addColumn(schema model.CIStr, tbl table.Table, spec *AlterSpecific
 	tb := tbl.(*tables.Table)
 	tb.Columns = newCols
 	// TODO: update index
-	// TODO: update default value
-	// update infomation schema
+	updateDefaultValue(ctx, tb, col)
```
Comment (qiuyesuifeng) [inline ddl/ddl.go]: Need check error.

File: ddl/ddl.go [hunk 2]
```diff
@@ -506,6 +507,44 @@ func (d *ddl) addColumn(schema model.CIStr, tbl table.Table, spec *AlterSpecific
...
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer it.Close()
+
+	prefix := t.KeyPrefix()
+	for it.Valid() && strings.HasPrefix(it.Key(), prefix) {
+		handle, err0 := util.DecodeHandleFromRowKey(it.Key())
+		if err0 != nil {
+			return errors.Trace(err)
```
Comment (qiuyesuifeng) [inline ddl/ddl.go]: errors.Trace(err0)?

File: ddl/ddl.go [hunk 3]
```diff
@@ -506,6 +507,44 @@ func (d *ddl) addColumn(schema model.CIStr, tbl table.Table, spec *AlterSpecific
...
+		}
+		k := t.RecordKey(handle, col)
+		colID, err0 := tables.ColumnID(k)
+		if err0 != nil {
+			return errors.Trace(err)
```
Comment (qiuyesuifeng) [inline ddl/ddl.go]: errors.Trace(err0)?

File: ddl/ddl.go [hunk 4]
```diff
@@ -506,6 +507,44 @@ func (d *ddl) addColumn(schema model.CIStr, tbl table.Table, spec *AlterSpecific
...
+		}
+		if colID != col.ID {
+			continue
+		}
+
+		types.Convert(col.DefaultValue, &col.FieldType)
+		t.SetColValue(txn, k, col.DefaultValue)
```
Comment (qiuyesuifeng) [inline ddl/ddl.go]: Need error check.

File: ddl/ddl.go [hunk 5]
```diff
@@ -506,6 +507,44 @@ func (d *ddl) addColumn(schema model.CIStr, tbl table.Table, spec *AlterSpecific
...
+		k := t.RecordKey(handle, col)
+		colID, err0 := tables.ColumnID(k)
+		if err0 != nil {
+			return errors.Trace(err)
+		}
+		if colID != col.ID {
+			continue
+		}
+
+		types.Convert(col.DefaultValue, &col.FieldType)
```
Comment (qiuyesuifeng) [inline ddl/ddl.go]: Seems this line is needless?

File: ddl/ddl_test.go [hunk 6]
```diff
@@ -78,13 +78,30 @@ func (ts *testSuite) TestT(c *C) {
 	tbStmt = statement("create table t2 (a int unique not null)").(*stmts.CreateTableStmt)
 	err = sessionctx.GetDomain(ctx).DDL().CreateTable(ctx, tbIdent2, tbStmt.Cols, tbStmt.Constraints)
 	c.Assert(err, IsNil)
+	tb, err := sessionctx.GetDomain(ctx).InfoSchema().TableByName(tbIdent2.Schema, tbIdent2.Name)
+	c.Assert(err, IsNil)
+	c.Assert(tb, NotNil)
+	_, err = tb.AddRecord(ctx, []interface{}{1})
```
Comment (qiuyesuifeng) [inline ddl/ddl_test.go]: Add `c.Assert(err, IsNil)`

File: table/tables/tables.go [hunk 7]
```diff
@@ -35,9 +35,13 @@ import (
 	"github.com/pingcap/tidb/sessionctx/variable"
 	"github.com/pingcap/tidb/table"
 	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/codec"
 	"github.com/pingcap/tidb/util/errors2"
 )
 
+// KeyLastItem is the last item in key.
+const KeyLastItem = 3
```
Comment (qiuyesuifeng) [inline table/tables/tables.go]: Seems no need to export?

Comment (qiuyesuifeng) [inline table/tables/tables.go]: And it is better to add some comment, like why here is 3.

File: ddl/ddl.go (hunk 5 above)
Comment (zimulala) [inline ddl/ddl.go]: got it.

File: ddl/ddl_test.go [hunk 8]
```diff
@@ -78,13 +78,30 @@ func (ts *testSuite) TestT(c *C) {
...
+	_, err = tb.AddRecord(ctx, []interface{}{2})
+	c.Assert(err, IsNil)
+	alterStmt := statement(`alter table t2 add b enum("bbb") first`).(*stmts.AlterTableStmt)
```
Comment (qiuyesuifeng) [inline ddl/ddl_test.go]: Should add a test that after u add a new column, u can select the updated value.

File: table/tables/tables.go [hunk 9]
```diff
@@ -263,21 +280,28 @@ func (t *Table) setOnUpdateData(ctx context.Context, touched []bool, data []inte
 	return nil
 }
 
+// SetColValue sets the column value.
+// If the column untouched, we don't need to do this.
+func (t *Table) SetColValue(txn kv.Transaction, key []byte, data interface{}) error {
+	v, err := t.EncodeValue(data)
+	if err != nil {
+		return err
```
Comment (qiuyesuifeng) [inline table/tables/tables.go]: New code use errors.Trace(err).

File: ddl/ddl_test.go (hunk 8 above)
Comment (zimulala) [inline ddl/ddl_test.go]: `tb.Row()` will get the updated values.

Comment (qiuyesuifeng) [inline ddl/ddl_test.go]: Yes, but need a value equal check.

File: table/tables/tables.go (hunk 7 above)
Comment (zimulala) [inline table/tables/tables.go]: got it.

File: ddl/ddl.go [hunk 10]
```diff
@@ -506,6 +508,47 @@ func (d *ddl) addColumn(schema model.CIStr, tbl table.Table, spec *AlterSpecific
...
+	defer it.Close()
+
+	prefix := t.KeyPrefix()
+	for it.Valid() && strings.HasPrefix(it.Key(), prefix) {
+		handle, err0 := util.DecodeHandleFromRowKey(it.Key())
+		if err0 != nil {
+			return errors.Trace(err0)
+		}
+		k := t.RecordKey(handle, col)
+		colID, err0 := tables.ColumnID(k)
```
Comment (qiuyesuifeng) [inline ddl/ddl.go]: Seems that in addColumn there will be no new column, so here `colID == col.ID` will always be true?

Comment (zimulala) [inline ddl/ddl.go]: yep.
here `colID == col.ID` is always true.  Because the key is produced by `RecordKey(handle, col)`。

File: ddl/ddl.go [hunk 11]
```diff
@@ -506,6 +508,47 @@ func (d *ddl) addColumn(schema model.CIStr, tbl table.Table, spec *AlterSpecific
 	return errors.Trace(err)
 }
 
+func updateDefaultValue(ctx context.Context, t *tables.Table, col *column.Col) error {
```
Comment (shenli) [inline ddl/ddl.go]: Any comment for this function? Or pick up a better name. "updateDefaultValue" is a little confused. "updateOldRows" may be better.

Comment (zimulala) [inline ddl/ddl.go]: got it.

**Comment by shenli** [comment]
LGTM

**Pull request merged**